EducationalTestingService/rsmtool
# Add support for cross-validation.

## Description

This PR adds support for cross-validation in RSMTool. 

- Add a new command-line utility called `rsmxval`. This utility basically works by calling `rsmtool`, `rsmeval`, and `rsmsummrize` in various ways. The detailed workflow is described at the top of `rsmxval.py` and also in the documentation.

- Since `rsmxval` calls these other tools as part of the new cross-validation workflow, we can't have all of those tools print their logs to the screen as they do now. We want those logs to go elsewhere. Therefore, these tools are modified to log to an optional, externally-provided logger instance. This also means changing several of the method types to be regular instance methods rather than classmethods or staticmethods.

- Add two new logging utilities: 

    + `get_file_logger()` creates a new file-based logger. This is used by `rsmxval` to create loggers for the `rsmtool`, `rsmeval`, and `rsmsummarize` calls that it makes. Since those invocations cannot log to the screen, they are made to log to files instead. Add a unit test for this utility under `test_utils.py:TestLogging` class.
    
    + `tqdm_joblib()` is a joblib-compatible context manager that can be used to run tasks in parallel and as each task completes, a tqdm progress bar updates. No test was added for this as it's not necessary.

- Add new cross-validation specific utility functions to `utils/cross_validation.py` and add unit tests for them under `test_utils.py:TestCrossValidation` class.

- Factor out the code that was used to write a dataframe to disk into a separate utility methods `DataWriter.write_frame_to_disk()` so that it can also be used by `rsmxval`. Also modify the frame writing so that it's correctly writing out jsonlines files. Add unit tests for this new method to `test_writer.py`.

- Add functional tests for `rsmxval` in the new file `test_experiment_rsmxval.py`. This file is set up to run as a sixth parallel job in Gitlab CI and Azure pipelines. Added new test utility functions `do_run_cross_validation()` and `check_run_cross_validation()` that are used by these functional tests.

- Allow `rsmxval` to generate configuration files automatically both in batch as well as interactive mode. Tweak `commandline.py` to make the help message for `rsmxval` a bit different than the others. Add tests for the `rsmxval` CLI to `test_cli.py`.

- Add comprehensive documentation for `rsmxval`. The built documentation for this branch can be found at [here](https://rsmtool.readthedocs.io/en/353-add-cross-validation/). This will be removed after this PR is merged.

- Modify Gitlab CI to use Python 3.8 instead of Python 3.7 and Modify Azure pipelines to use Python 3.9 instead of Python 3.8. This is because we are going to drop support for Python 3.7 as part of this version.

- Tweak the `check_scaled_coefficients()` test utility function to take the output directory as an argument instead of taking an experiment name. This allows it to be used for `rsmxval` functional tests too.

- Add `tqdm` to `requirements.txt`.

- Clean up imports and fix typos. Remove unnecessary executable flag.

## How to review

- Ensure that the cross-validation workflows makes sense and is useful.
- Review all code and tests and look for bugs and other issues (readability, typos etc.)
- Read the documentation and check that it makes sense, isn't missing anything, and has no types
  or other issues.
- Try out the tutorial too.
- Run several cross-validation experiments with various data files and various config options.
- Test automated configuration generation for `rsmxval` - both batch and interactive and run
  those automatically generated config files to make sure they work.
